test/services/database/challenge_repository_test.dart

```diff
@@ -0,0 +1,206 @@
+
+Future<List<PostData>> addPosts(
+  PostRepositoryService postRepository,
+  GeoPoint pos,
+  int n,
+) async {
+  final fakePosts = PostDataGenerator.generatePostData(n);
+  for (final post in fakePosts) {
+    await postRepository.addPost(post, pos);
+  }
+  return fakePosts;
+}
+
+void main() {
+  late FakeFirebaseFirestore firestore;
+  late PostRepositoryService postRepository;
+  late ChallengeRepositoryService challengeRepository;
+
+  setUp(() async {
+    firestore = FakeFirebaseFirestore();
+    postRepository = PostRepositoryService(firestore: firestore);
+    challengeRepository = ChallengeRepositoryService(
+      firestore: firestore,
+      postRepositoryService: postRepository,
+    );
+  });
+
+  const userPos = userPosition1;
+  final inChallengeRange = GeoPointGenerator.createOnEdgeInsidePosition(
+    userPos,
+    ChallengeRepositoryService.maxChallengeRadius,
+  );
+
+  final uid = testingUserFirestoreId;
+
+  group("ChallengeRepositoryService", () {
+    test("Get new challenges", () async {
+      final fakePosts = await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      expect(challenges.length, ChallengeRepositoryService.maxActiveChallenges);
+      for (final challenge in challenges) {
+        expect(challenge.data.isCompleted, false);
+        expect(challenge.data.expiresOn.compareTo(Timestamp.now()), isPositive);
+
+        final actualPost = await postRepository.getPost(challenge.postId);
+        expect(fakePosts.contains(actualPost.data), true);
+      }
+    });
+
+    test("Complete a challenge", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+      final challenges = await challengeRepository.getChallenges(uid, userPos);
+
+      final challenge = challenges.first;
+      expect(challenge.data.isCompleted, false);
+      await challengeRepository.completeChallenge(
+        uid,
+        challenge.postId,
+      );
+
+      final updatedChallenges =
+          await challengeRepository.getChallenges(uid, userPos);
+      expect(updatedChallenges.length, 1);
+      expect(updatedChallenges.first.data.isCompleted, true);
+    });
+
+    test("Multiple gets with only one challenge available", () async {
+      await addPosts(postRepository, inChallengeRange, 1);
+
+      for (int i = 0; i < 10; i++) {
+        final challenges =
+            await challengeRepository.getChallenges(uid, userPos);
+
+        expect(challenges.length, 1);
+      }
+    });
+
+    test("Challenge posts are unique", () async {
+      await addPosts(
+        postRepository,
+        inChallengeRange,
+        ChallengeRepositoryService.maxActiveChallenges,
+      );
+
+      final challenges = await challengeRepository.getChallenges(
+        uid,
+        userPos,
+      );
+      final postIds = challenges.map((e) => e.postId).toSet();
```

Comment:
To complete the triforce, my turn do make this suggestion! :D
```suggestion
      final postIds = challenges.map((challenge) => challenge.postId).toSet();
```